tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,786 @@
+		}
+		// Verify magic number.
+		if m := binary.BigEndian.Uint32(b.Range(0, MagicChunksSize)); m != MagicHeadChunks {
+			return errors.Errorf("invalid magic number %x", m)
+		}
+
+		// Verify chunk format version.
+		if v := int(b.Range(MagicChunksSize, MagicChunksSize+ChunksFormatVersionSize)[0]); v != chunksFormatV1 {
+			return errors.Errorf("invalid chunk format version %d", v)
+		}
+
+		maxt := binary.BigEndian.Uint64(b.Range(HeaderMaxtOffset, HeaderMaxtOffset+MintMaxtSize))
+		if maxt == 0 {
+			// This is possible if Prometheus crashes and the maxt was unwritten to the
+			// last head chunk file. As a safe buffer, we set the maxt to mint + 1.5 times head chunk file time range.
+			f, err := os.OpenFile(files[i], os.O_WRONLY|os.O_CREATE, 0666)
+			if err != nil {
+				return err
+			}
+			defer func() {
+				var merr tsdb_errors.MultiError
+				merr.Add(returnErr)
+				merr.Add(f.Close())
+				returnErr = merr.Err()
+			}()
+			mint := binary.BigEndian.Uint64(b.Range(HeaderMintOffset, HeaderMintOffset+MintMaxtSize))
+			binary.BigEndian.PutUint64(cdm.writeBuf[:], mint+(uint64(cdm.maxFileMs)*3/2))
+			if _, err := f.WriteAt(cdm.writeBuf[:MintMaxtSize], HeaderMaxtOffset); err != nil {
+				return err
+			}
+		}
+
+		cdm.size += int64(b.Len())
+	}
+
+	return nil
+}
+
+func listChunkFiles(dir string) (map[int]string, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	res := map[int]string{}
+	for _, fi := range files {
+		seq, err := strconv.ParseUint(fi.Name(), 10, 64)
+		if err != nil {
+			continue
+		}
+		res[int(seq)] = filepath.Join(dir, fi.Name())
+	}
+	return res, nil
+}
+
+// WriteChunk writes the chunk to the disk.
+// The returned chunk ref is the reference from where the chunk encoding starts for the chunk.
+func (cdm *ChunkDiskMapper) WriteChunk(seriesRef uint64, mint, maxt int64, chk chunkenc.Chunk) (chkRef uint64, err error) {
+	cdm.writePathMtx.Lock()
+	defer cdm.writePathMtx.Unlock()
+
+	if cdm.closed {
+		return 0, ErrChunkDiskMapperClosed
+	}
+
+	if cdm.shouldCutNewFile(len(chk.Bytes()), maxt) {
+		if err := cdm.cut(mint); err != nil {
+			return 0, err
+		}
+	}
+
+	// if len(chk.Bytes())+MaxHeadChunkMetaSize >= writeBufferSize, it means that chunk >= the buffer size;
+	// so no need to flush here, as we have to flush at the end (to not keep partial chunks in buffer).
+	if len(chk.Bytes())+MaxHeadChunkMetaSize < writeBufferSize && cdm.wbuf.Available() < MaxHeadChunkMetaSize+len(chk.Bytes()) {
+		if err := cdm.flushBuffer(); err != nil {
+			return 0, err
+		}
+	}
+
+	cdm.crc32.Reset()
+
+	// The upper 4 bytes are for the head chunk file index and
+	// the lower 4 bytes are for the head chunk file offset where to start reading this chunk.
+	chkRef = chunkRef(uint64(cdm.curFileSequence), uint64(cdm.n))
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], seriesRef)
+	if err := cdm.writeAndAppendToCRC32(cdm.writeBuf[:SeriesRefSize]); err != nil {
```

Comment:
why not make the buf big enough to hold all the chunk metadata, and than you can write to the buffer only 2 times 1 for the metadata and one for the chunks data? I mean why do you have to write to buffer for every chunk element?

Reply:
Sounds fair and that will clean-up the code too. Max meta size is not big enough to be a concern. I will change it.